plugin/template/setup.go

```diff
@@ -0,0 +1,143 @@
+package template
+
+import (
+	"regexp"
+	gotmpl "text/template"
+
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+
+	"github.com/mholt/caddy"
+	"github.com/miekg/dns"
+)
+
+func init() {
+	caddy.RegisterPlugin("template", caddy.Plugin{
+		ServerType: "dns",
+		Action:     setupTemplate,
+	})
+}
+
+func setupTemplate(c *caddy.Controller) error {
+	templates, err := templateParse(c)
+	if err != nil {
+		return plugin.Error("template", err)
+	}
+
+	c.OnStartup(OnStartupMetrics)
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		return Handler{Next: next, Templates: templates}
+	})
+
+	return nil
+}
+
+func templateParse(c *caddy.Controller) (templates []template, err error) {
+	templates = make([]template, 0)
+
+	for c.Next() {
+		t := template{}
+		if !c.NextArg() {
+			return nil, c.ArgErr()
+		}
+
+		class, ok := dns.StringToClass[c.Val()]
```

Comment:
As noted in https://github.com/miekg/dns/issues/577#issuecomment-346235028 , "ANY" is not a real class mnemonic. This should be prepared for its removal, e.g.
```go
if c.Val() == "ANY" {
	t.class = dns.ClassANY
} else {
	class, ok := dns.StringToClass[c.Val()]
	if !ok {
		return nil, c.Errf("invalid query class %s", c.Val())
	}
	t.class = class
}
```

Reply:
The linked page lists ANY: `255 QCLASS * (ANY)`
https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-2

ANY is not ALL, so you are free to respond with nothing.

Reply:
Ah, removing from the list, yeah, that's fine, we just have to handle the ordering of these 2 PRs

Reply:
btw, ANY is also an invalid RTYPE mnemonic so you'll want similar code for that.

Reply:
It is valid for querying and matching. It's '*' or 'ANY'.
It is not valid in DNS responses, **but these values here are never used in RRs, they are used for matching**.

Reply:
I feel like you're misunderstanding me... ANY is not valid for either class or type in an RFC 1035 master file, even though the values are assigned and CLASS255 or TYPE255 _would_ be valid (if nonsensical). You're relying on an entry in the `StringToClass` map that doesn't belong there, and you should stop doing so in order to avoid breaking if the bug in the base library is fixed.

Reply:
Yes, I totally understand that, but they are not used in the context of 1035 here. They are not even used in the context of an RR here.

They are used to *match incoming types*. The same way you can ask a server to return *something* by sending an ANY query.

Reply:
I'm not telling you to drop "ANY" from your matchers, I'm telling you to stop assuming that "ANY" will be in `StringToClass` (which _is_ used for parsing presentation format).

Reply:
Which is your suggestion in another PR, I know that.

But I could also invert that: Why not fix the parsing to only use mnemonic of that context?

The documents you pull as reference call it ANY. It is commonly referred to as ANY. It is just not valid in a certain, different, context.

Anyway, fixing this either way is trivial. I don't see any value in the debate here.